core/server/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -632,55 +874,129 @@ public boolean equals(Object o) {
-      if (pathComponents[0].equals("")) {
-        inodes.add(mRoot);
-        return TraversalResult.createFoundResult(mRoot, nonPersistedInodes, inodes);
+      if (getLockModeForComponent(0, pathComponents.length, lockMode, lockHints) == LockMode.READ) {
+        lockGroup.lockRead(mRoot);
       } else {
-        throw new InvalidPathException("File name starts with " + pathComponents[0]);
+        lockGroup.lockWrite(mRoot);
+      }
+      inodes.add(mRoot);
+      TraversalResult result =
+          traverseToInodeInternal(pathComponents, inodes, nonPersistedInodes, lockGroup, lockMode,
+              lockHints);
+      valid = true;
+      return result;
+    } finally {
+      if (!valid) {
+        lockGroup.close();
       }
     }
+  }
 
-    Inode<?> current = mRoot;
-    inodes.add(current);
+  private TraversalResult traverseToInode(InodePath inodePath, LockMode lockMode)
+      throws InvalidPathException {
+    // the inodePath is guaranteed to already include at least the root inode.
+    if (!(inodePath instanceof ExtensibleInodePath)) {
+      throw new InvalidPathException("Not an ExtensibleInodePath: " + inodePath.getUri());
+    }
+    ExtensibleInodePath extensibleInodePath = (ExtensibleInodePath) inodePath;
+    List<Inode<?>> inodes = extensibleInodePath.getInodes();
+    InodeLockGroup lockGroup = extensibleInodePath.getLockGroup();
+    List<Inode<?>> nonPersistedInodes = Lists.newArrayList();
+    for (Inode<?> inode : inodes) {
+      if (!inode.isPersisted()) {
+        nonPersistedInodes.add(inode);
+      }
+    }
+    return traverseToInodeInternal(extensibleInodePath.getPathComponents(), inodes,
+        nonPersistedInodes, lockGroup, lockMode, null);
+  }
 
-    // iterate from 1, because 0 is root and it's already added
-    for (int i = 1; i < pathComponents.length; i++) {
+  private TraversalResult traverseToInodeInternal(String[] pathComponents, List<Inode<?>> inodes,
+      List<Inode<?>> nonPersistedInodes, InodeLockGroup lockGroup, LockMode lockMode,
+      List<LockMode> lockHints)
+      throws InvalidPathException {
+    Inode<?> current = inodes.get(inodes.size() - 1);
+    for (int i = inodes.size(); i < pathComponents.length; i++) {
       Inode<?> next = ((InodeDirectory) current).getChild(pathComponents[i]);
       if (next == null) {
-        // The user might want to create the nonexistent directories, so return the traversal result
-        // current inode with the last Inode taken, and the index of the first path component that
-        // couldn't be found.
-        return TraversalResult.createNotFoundResult(current, i, nonPersistedInodes, inodes);
-      } else if (next.isFile()) {
+        if (lockMode != LockMode.READ
+            && getLockModeForComponent(i - 1, pathComponents.length, lockMode, lockHints)
+            == LockMode.READ) {
+          // The lock mode is one of the WRITE modes, but the previous inode was READ locked.
+          lockGroup.unlockPrevious();
```

Comment:
ah, I got confused by the use of previous and current here (they both refer to the same inode) ... I think it would make sense to rename unlockPrevious to unlockLast?